### PR TITLE
Improved argument description for parameters originating from tuples

### LINKED ISSUE #26563: Better argument description for inferred argument types
## Search Terms

tsserver, server, infer, argument help

## Suggestion

Get a better argument description when accessing the argument help of an inferred argument list.

When accessing the argument information for the following code ([Playground](https://www.typescriptlang.org/play/index.html#src=type%20Parameters%3CT%3E%20%3D%20T%20extends%20(...args%3A%20infer%20U)%20%3D%3E%20any%20%3F%20U%20%3A%20never%3B%0Adeclare%20function%20fun1(x%3A%20number%2C%20y%3A%20string)%3A%20number%3B%0Adeclare%20function%20fun2(...args%3A%20Parameters%3Ctypeof%20fun1%3E)%3A%20string%3B%0A%0Afun2(%2F*%20get%20parameter%20help%20here%20*%2F)%3B%0A%0Afun2(10%2C%20%22text%22)%3B%0A%2F*%20%5E%20hover%20here%20*%2F)), both vscode and the playground only show `fun2(...args: [number, string]): string`:

```ts
type Parameters<T> = T extends (...args: infer U) => any ? U : never;
declare function fun1(x: number, y: string): number;
declare function fun2(...args: Parameters<typeof fun1>): string;

fun2(/* get parameter help here */);

fun2(10, "text");
/* ^ hover here */
```

It'd be a lot more intuitive, if the help would resemble the actual parameters.

When hovering over `fun2` in the playground, it even shows the more readable information `fun2(x: number, y: string): number`, so this might be an easy improvement.

## Use Cases

Even though the current display is perfectly valid, it makes it hard to understand what the parameters were meant for, so the only use case for this would be readability/UX.

## Related Issues

Not exactly a duplicate of #26132 (which seems fixed in 3.1.0-dev.20180818), but also not a totally different one. If you consider this feature request a duplicate, feel free to close it.

## Checklist

My suggestion meets these guidelines:
* [x] This wouldn't be a breaking change in existing TypeScript / JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. new expression-level syntax)

### PR DESCRIPTION
**Pull request by dragomirtitian**
Improved argument description for parameters originating from tuples that were extracted from functions. (ie mapped functions).

If the function parameters come from a tuple extracted from function parameters, then the signature help now provides the same information as the quick info. This means that if the tuple was extracted from a function we will now have the same names for the parameters as in the original function.

Ex:

```ts
function fnTest(str: string, num: number) { }
declare function wrap<A extends any[], R>(fn: (...a: A) => R) : (...a: A) => R;
var fnWrapped = wrap(fnTest);
fnTest(/*1*/) // At 1 - fnWrapped(str: string, num: number): void instead of function fnTest(...a: [string, number]) : void
```

Implementation notes:

1. I added `getExpandedParameters` to `TypeChecker` but marked it as internal so as not to impact the public API. This function might however be useful for clients of the API. Let me know if I should remove the internal.

2.  If the argument of the mapped function is a single rest parameter the name of the parameter is not preserved in the quick info. Since this fix uses the same function as quick info (`getExpandedParameters`) it will suffer from the same issue. 

```ts
function fnTestVariadic (...num: number[]) { }
var fnVariadicWrapped = wrap(fnTestVariadic);
fnVariadicWrapped() // fnVariadicWrapped(...a: number[]): void instead of fnVariadicWrapped(...num: number[]): void
```

This seems to be caused by the fact that the signature parameters will just have the rest parameter (`...a: num[]`) not a tuple with the rest (`...a: [...number[]]`).

This could be the topic of a different issue.

3.  If the parameters do not originate from a function (ex: `function fnTest3(...a: [number, string]) { }`) the quick info shows a signature with the tuple spread to several parameters and an index added (`function fnTest3(a_0: number, a_1: string): void` ) this will also now be shown by the signature help as well. While this does not seem like a great improvement at least it is consistent.

Fixes #26563, #28127, #29514

### CODE REVIEW

**Comment by dragomirtitian** [comment]
On a personal note, this issue has been bugging me since 3.0, hope it gets merged :)

**Comment by dragomirtitian** [comment]
@RyanCavanaugh Is there anything blocking this PR and anything I can do to move it along?  Or is it just a matter of time and priorities and someone will get to it when they get to it :) ?

File: src/compiler/types.ts [hunk 1]
```diff
@@ -3032,6 +3032,7 @@ namespace ts {
         /* @internal */
         getPromisedTypeOfPromise(promise: Type, errorNode?: Node): Type | undefined;
         getReturnTypeOfSignature(signature: Signature): Type;
+        /* @internal */ getExpandedParameters(sig: Signature): ReadonlyArray<Symbol>;
```
Comment (ahejlsberg) [inline src/compiler/types.ts]: Also add `hasEffectiveRestParameter` as an `@internal` callback.

File: src/services/signatureHelp.ts [hunk 2]
```diff
@@ -566,15 +566,24 @@ namespace ts.SignatureHelp {
...
+        const expandedParameters = checker.getExpandedParameters(candidateSignature);
+        const parameters = expandedParameters.map(p => createSignatureHelpParameterForParameter(p, checker, enclosingDeclaration, sourceFile, printer));
+
+        // If parameters are not the same as the expanded parameters, we need to reevaluate whether this a variadic signature based on the last parameter
+        if (expandedParameters !== candidateSignature.parameters) {
+            const restCandidate = lastOrUndefined(expandedParameters);
+            isVariadic = !!restCandidate && !!(getCheckFlags(restCandidate) & CheckFlags.RestParameter);
+        }
+
         return { isVariadic, parameters, prefix: [...typeParameterParts, punctuationPart(SyntaxKind.OpenParenToken)], suffix: [punctuationPart(SyntaxKind.CloseParenToken)] };
```
Comment (ahejlsberg) [inline src/services/signatureHelp.ts]: Compute `isVariadic` using `hasEffectiveRestParameter(candidateSignature)`. No need for special handling of the expanded case.

**Comment by ahejlsberg** [comment]
@dragomirtitian Added a few comments in the code. If you update the PR we'll get it merged.

File: src/services/signatureHelp.ts (hunk 2 above)
Comment (dragomirtitian) [reply src/services/signatureHelp.ts]: @ahejlsberg I missed `hasEffectiveRestParameter` when digging around in the code. Made the change, 10x.

**Pull request merged**